### ISSUE TITLE
xbps-rindex -a: validate deps

### ISSUE THREAD

**Issue by xtraeme** [void-linux/xbps#259]
I'm working in a feature that I had in my mind for a while.

`xbps-rindex -a` to add a package to the repo index, does not validate dependencies, thus a package with unresolved dependencies can be added and nobody would notice until it's installed.

This feature will modify `xbps-rindex -a` to always validate its required dependencies before the pkg is registered into the repository index.

**Comment by st3r4g** [comment]
How would it work when the dependency is in another repo (e.g. `-dbg` packages)?

**Comment by xtraeme** [comment]
via rpool

**Comment by xtraeme** [comment]
the same strategy will have to be made for `xbps-rindex -c`. While cleaning up the repo index, before unindexing an entry it will have to verify its revdeps.

**Comment by xtraeme** [comment]
this is the theory, let's see if I can come up with something that works :-)

**Comment by xtraeme** [comment]
another interesting feature would be adding a new mode that does the validation for existing repos.

**Comment by xtraeme** [comment]
ok found an issue in the first implementation, it just needs to verify all pkgs at once before flushing repodata :-) otherwise if you do `xbps-rindex -a repo/foo.xbps` its dependencies might not be there...

**Comment by xtraeme** [comment]
thinking about it, this should be enabled via a new flag. Because while building everything from scratch the validation will be too strict.

**Comment by xtraeme** [comment]
hmm this is not that easy, but I still think it's doable

**Comment by Duncaen** [comment]
I think this would be a nice opportunity to introduce a separate tool.

* `xbps-rindex` is the tool for `xbps-src` and local repositories (always staged).
* `xbps-repoadmin` is responsible for everything related to publishing repositories.

## `xbps-rindex`

* adds packages to the architectures `-stagedata`
* remove signing staging etc.

## `xbps-repoadmin`/`xbps-repodb`?

* works on multiple `-stagedata` files.
* checks shlibs and dependencies and then selectively moves satisfied packages into repodata.
* this tool is also responsible for signing it will always be executed with the signing keys available.
* if it moves a package from `-stagedata` to `-repodata` it signs them.
* in the future this tool will also be able to sign the `-repodata` files whenever it changes them.
* this tool also does the cleanup stuff, it knows multiple repositories, knows `-repodata` and `-stagedata` and can do a better job at cleaning them up.

* Optionally also work on multiple architectures, this would help to solve `noarch` issues if we want to keep them. But in general `noarch` usually depend on other packages that are not `noarch` and therefor can introduce uninstallable packages.

**Comment by xtraeme** [comment]
omg found the issue in `repodata_commit` it returns a bool but later on uses int... got this almost working :-)

**Comment by xtraeme** [comment]
xbps-rindex/add_test:add_validate_deps  ->  passed  [0.082s]

**Comment by xtraeme** [comment]
bad news is that there 6 test cases now broken :D

**Comment by xtraeme** [comment]
@Duncaen sounds good, but one thing at a time. Once I got this working we can start with repoadmin.

**Comment by xtraeme** [comment]
can't believe this bool vs int confusion took me 1h until I figured it out rofl

**Comment by xtraeme** [comment]
I added `--validate` (long option) to choose this behaviour.

**Comment by xtraeme** [comment]
yay, it works!

```
xbps-rindex/add_test:add_validate_deps_multirepo  ->  passed  [0.088s]
xbps-rindex/add_test:add_validate_deps  ->  passed  [0.105s]
```

**Comment by xtraeme** [comment]
so `xbps-rindex` now accepts `-C/--confdir` and `--validate`, previous `-C` is now `-H`.

**Comment by xtraeme** [comment]
`-C/--confdir` is so important because this is where repositories should be declared, like the other xbps utils.

**Comment by xtraeme** [comment]
Ok preparing PR for review/comments :-)

**Comment by Duncaen** [comment]
Some flags I think `xbps-repoadmin` should have

* `-i --ignore-conf-repos`
* `--repository=<path>`
* ` -C --config <dir>`

So like the proposed `xbps-rindex` change it would work on the given rpool, but it would be able to move packages from each of those repositories staging data to their corresponding -repodata after verifying shlibs and dependencies and probably reverse dependencies.

**Comment by xtraeme** [comment]
@Duncaen right. repoadmin will do shlib checking and validate deps/revdeps in insertions/deletions.

**Comment by xtraeme** [comment]
and once all checks were ok, sign, move to repodata and be done!

**Comment by xtraeme** [comment]
BTW I don't like the `repoadmin` name... how about `repopush` or `rpush`?

**Comment by Duncaen** [comment]
I think "push" sounds a bit limited.

Maybe `xbps-repodb` as counterpart to `xbps-pkgdb`.

I thought about `xbps-repodb` and `xbps-repoadmin` because besides moving packages between staging, it could also do the cleanup tasks because it would be aware of -stagedata and -repodata.
So `xbps-rindex` becomes a tool to really just "index" packages and `xbps-repoadmin`/`xbps-repodb` (or some other name) is something that is a tool to manage and administrate public repositories.

**Comment by xtraeme** [comment]
I'll repodb then

**Comment by Duncaen** [comment]
Another feature I was thinking about is not using `XBPS_ARCH`/`XBPS_TARGET_ARCH` and instead use maybe a architecture flag that has comma/space separated architectures or a architecture flag that can be repeated.
If `xbps-repodb` can also act on repositories for multiple architectures, we can "solve" the noarch problems. It would be aware of noarch packages for each repository and then can safely delete noarch packages without having them to add to each repository index.

**Comment by xtraeme** [comment]
`-a foo` like xbps-src, I agree.

**Comment by xtraeme** [comment]
@Duncaen just to be clear, new xbps-rindex should only register pkgs into stagedata, isn't it?

**Comment by Duncaen** [comment]
Yes, this is the idea, for local repos/xbps-src everything stays basically the same.
The good stuff comes with the next step when you publish a repository.

**Comment by xtraeme** [comment]
ok I think I got it now. we will need another build step in buildbot after xbps-src finishes, that execs xbps-repodb.

**Comment by xtraeme** [comment]
in that case I think new xbps-rindex can be simplified as:

```
$ xbps-rindex [-f] /path/to/repo/*.xbps
```
